### PR TITLE
Configure API Key Middleware

### PR DESCRIPTION
**Pull request by JohnathonBowers**
## Description

These changes implement middleware for validating API keys sent in request headers coming from `OpenSearch_Parse_Changelog_Action` workflow runs.

## Features

In order to secure sensitive information and ensure proper error handling, this PR:

- creates a `validateChangelogPRBridgeSecretKey` middleware function, which (1) checks to make sure that an API key has been configured in the deployed instance of the Express app, (2) compares the configured API key (`authorizedApiKey`) with the key in the request header (`receivedApiKey`) to make sure that a matching key is present, and (3) either throws errors as necessary or calls the next middleware function

- configures the `validateChangelogPRBridgeSecretKey` middleware function in the Express application flow in `main.js`. The middleware applies to all routes and is invoked immediately after the request body is parsed so that no further action is taken if an API key is missing or invalid.

- creates a new `errors` directory for modularizing the construction of custom error classes. A base `CustomError` class is declared in `base.errors.js`. The error classes pertinent to API key validation extend this `CustomError` class and are declared in `auth.errors.js`. They distinguish between messages that are intended for server-side debugging (`internalMessage`) and general messages to be sent back on the response object (`publicMessage`).

- adds an entry in the `.env.sample` file to prompt contributors to configure their own keys when developing in their local environments.

## Testing

These changes have been tested manually in my local environment to make sure that the middleware and error handling work as expected.

**NOTE:** In order for this middleware to work properly, an environment variable must be configured in the deployed instance of the Express application. This variable must be named `CHANGELOG_PR_BRIDGE_SECRET_KEY`, and the value must be identical to what is set as a repository secret in the *base repository* that uses the `OpenSearch_Parse_Changelog_Action` in a workflow run.

## Issues

Resolves Issue #2

### CODE REVIEW

File: src/middlewares/auth.middlewares.js [hunk 1]
```diff
@@ -1,4 +1,6 @@
 import { authServices } from "../services/index.js"; // Adjust the path as necessary
+import { CHANGELOG_PR_BRIDGE_SECRET_KEY } from "../config/constants.js";
+import { AuthorizedAPIKeyNotConfiguredError, UnauthorizedAPIKeyReceivedError } from "../errors/index.js";
```
Comment (BigSamu) [inline src/middlewares/auth.middlewares.js:3]: Just import `UnauthorizedApiKeyError `and use it in the code

File: src/middlewares/auth.middlewares.js [hunk 2]
```diff
@@ -32,3 +34,18 @@ export async function ensureGitHubAppInstalled(req, res, next) {
...
   }
 }
+
+export const validateChangelogPRBridgeSecretKey = (req, res, next) => {
+  const receivedApiKey = req.headers['x-api-key'];
+  const authorizedApiKey = CHANGELOG_PR_BRIDGE_SECRET_KEY;
+
+  if (!authorizedApiKey || authorizedApiKey.trim() === '') {
+    throw new AuthorizedAPIKeyNotConfiguredError();
+  }
```
Comment (BigSamu) [inline src/middlewares/auth.middlewares.js:44]: Remove this

```suggestion
```

File: src/middlewares/error.middlewares.js [hunk 3]
```diff
@@ -1,10 +1,10 @@
-
 export const errorRequestHandler = (err, req, res, next) => {
-  console.error(err.stack);
+  // Log internal message for server-side debugging
+  console.error(err.internalMessage || err.stack);
```
Comment (BigSamu) [inline src/middlewares/error.middlewares.js:3]:
```suggestion
  console.error(err.stack);
```

File: src/middlewares/error.middlewares.js [hunk 4]
```diff
@@ -1,10 +1,10 @@
...
 
   // Set the response status code and message
   const statusCode =  err.status || 500;
-  const message = err.message || "Internal Server Error";
+  const message = err.publicMessage || err.message || "Internal Server Error";
```
Comment (BigSamu) [inline src/middlewares/error.middlewares.js:7]:
```suggestion
  const message = err.message || "Internal Server Error";
```

File: src/main.js [hunk 5]
```diff
@@ -7,6 +7,7 @@ import fileRouter from "./routes/file.routes.js";
 import {
   errorRequestHandler,
   ensureGitHubAppInstalled,
+  validateChangelogPRBridgeSecretKey,
```
Comment (BigSamu) [inline src/main.js:10]: I would change the name of the function from `validateChangelogPRBridgeSecretKey` to `verifyAuthorizationApiRequest` for better consistency.

File: src/middlewares/auth.middlewares.js [hunk 6]
```diff
@@ -32,3 +34,18 @@ export async function ensureGitHubAppInstalled(req, res, next) {
     res.status(500).json({ message: "Internal Server Error" });
   }
 }
+
+export const validateChangelogPRBridgeSecretKey = (req, res, next) => {
+  const receivedApiKey = req.headers['x-api-key'];
+  const authorizedApiKey = CHANGELOG_PR_BRIDGE_SECRET_KEY;
```
Comment (BigSamu) [inline src/middlewares/auth.middlewares.js:40]: avoid DRY

```suggestion
```

File: src/middlewares/auth.middlewares.js [hunk 7]
```diff
@@ -32,3 +34,18 @@ export async function ensureGitHubAppInstalled(req, res, next) {
...
+
+  if (!authorizedApiKey || authorizedApiKey.trim() === '') {
+    throw new AuthorizedAPIKeyNotConfiguredError();
+  }
+
+  if (!receivedApiKey || receivedApiKey !== authorizedApiKey) {
+    throw new UnauthorizedAPIKeyReceivedError();
+  }
```
Comment (BigSamu) [inline src/middlewares/auth.middlewares.js:48]:
```suggestion
  if (!receivedApiKey || receivedApiKey !== CHANGELOG_PR_BRIDGE_SECRET_KEY) {
    throw new UnauthorizedApiKeyError();
  }
```

**Review by BigSamu — CHANGES_REQUESTED** [review]

File: src/errors/base.errors.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+     * Creates a new CustomError instance.
+     * @param {number} status - The HTTP status code.
+     * @param {string} internalMessage - The internal error message to be logged for server-side debugging.
+     * @param {string} publicMessage - The public error message to be returned in the response.
+     */
+    constructor(status, internalMessage = "An internal error occurred", publicMessage = "Internal Server Error") {
+        super(internalMessage);
+        this.status = status;
+        this.internalMessage = internalMessage;
+        this.publicMessage = publicMessage;
```
[the remark this answers is not in the document]
Comment (JohnathonBowers) [reply src/errors/base.errors.js:15]: Sounds good. I'll make this change.

File: src/errors/auth.errors.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    /**
+     * Creates a new UnauthorizedAPIKeyReceivedError instance.
+     */
+    constructor() {
+        super(
+            401,
+            `Unauthorized API key received on ${new Date().toISOString()}`,
+            "Unauthorized access"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (JohnathonBowers) [reply src/errors/auth.errors.js:32]: Sounds good. I've removed the `AuthorizedAPIKeyNotConfiguredError` and changed the name of `UnauthorizedAPIKeyReceivedError` to `UnauthorizedAPIKeyError`.

File: src/middlewares/auth.middlewares.js (hunk 6 above)
Comment (JohnathonBowers) [reply src/middlewares/auth.middlewares.js:40]: I'm not sure how the code is repeating itself here. The `receivedApiKey` is the key coming in on the request header. The `authorizedApiKey` is the key stored on the server that the `receivedApiKey` will be checked against. We need both of these variables for the authentication to take place.

Comment (JohnathonBowers) [reply src/middlewares/auth.middlewares.js:40]: The reason I stored `CHANGELOG_PR_BRIDGE_SECRET_KEY` in a different variable is to make the code more readable. These variables make it easier to understand the source of each key.

File: src/middlewares/auth.middlewares.js (hunk 7 above)
Comment (JohnathonBowers) [reply src/middlewares/auth.middlewares.js:48]: Simplifying this to:

```
if (receivedApiKey !== authorizedApiKey) {
    throw new UnauthorizedApiKeyError();
}
```

If there is no API key coming in on the request header, then, by necessity, it won't match the API key stored on the server. So, the additional check for the existince of an API key on the request header is unnecessary.

File: src/middlewares/auth.middlewares.js (hunk 2 above)
Comment (JohnathonBowers) [reply src/middlewares/auth.middlewares.js:44]: Removed.

File: src/main.js (hunk 5 above)
Comment (JohnathonBowers) [reply src/main.js:10]: To keep consistent with the way the incoming API key is named in the middleware function, I'll name this function `verifyReceivedApiKey` for now. We can change it later if we feel there is a better name.

File: src/middlewares/error.middlewares.js (hunk 3 above)
Comment (JohnathonBowers) [reply src/middlewares/error.middlewares.js:3]: Change made.

File: src/middlewares/error.middlewares.js (hunk 4 above)
Comment (JohnathonBowers) [reply src/middlewares/error.middlewares.js:7]: Changed.

File: src/main.js (hunk 5 above)
Comment (BigSamu) [reply src/main.js]: Ok perfect!

File: src/middlewares/auth.middlewares.js (hunk 6 above)
Comment (BigSamu) [reply src/middlewares/auth.middlewares.js]: Mmmm, you are repeating yourself here. I prefer to be concise a have the fewer lines possible. Indeed the constant `CHANGELOG_PR_BRIDGE_SECRET_KEY` we change it to `CHANGELOG_PR_BRIDGE_API_KEY`, so this is enough for readability.  

Anyway your call on this. Don't want to be a stopper.

File: src/middlewares/auth.middlewares.js (hunk 7 above)
Comment (BigSamu) [reply src/middlewares/auth.middlewares.js]: I would say

```js
if (receivedApiKey !== CHANGELOG_PR_BRIDGE_API_KEY) {
    throw new UnauthorizedApiKeyError();
}
```

But once again, your call on this

**Review by BigSamu — APPROVED** [review]

File: src/middlewares/auth.middlewares.js (hunk 7 above)
Comment (JohnathonBowers) [reply src/middlewares/auth.middlewares.js]: Good point. I had thought about that possibility after making this comment. I agree we should include the check for the presence of the receivedApiKey in the function.

**Pull request merged**